File: solutions/java/triangle/1/src/main/java/Triangle.java
import java.util.HashSet;
import java.util.Set;

class Triangle {

    double x;
    double y;
    double z;

    int sideSize;

    Triangle(double x, double y, double z) throws TriangleException {
        this.x = x;
        this.y = y;
        this.z = z;
        validate();
        Set<Double> set = new HashSet<>() {{ add(x); add(y); add(z); }};
        sideSize = set.size();
    }

    boolean isEquilateral() {
        return sideSize == 1;
    }

    boolean isIsosceles() {
        return sideSize <= 2;
    }

    boolean isScalene() {
        return sideSize == 3;
    }

    void validate() throws TriangleException {
        if (!(x * y * z != 0.0
                && x + y > z
                && x + z > y
                && y + z > x
        )) throw new TriangleException();
    }
}

File: solutions/kotlin/triangle/2/src/main/kotlin/Triangle.kt
import java.lang.IllegalArgumentException

class Triangle<T : Number>(a: T, b: T, c: T) {
    private val sides = listOf(a.toDouble(), b.toDouble(), c.toDouble()).sorted()
    private val sideSize = sides.toSet().size

    val isEquilateral: Boolean = isValid() && sideSize == 1
    val isIsosceles: Boolean = isValid() && sideSize <= 2
    val isScalene: Boolean = isValid() && sideSize == 3

    private fun isValid(): Boolean {
        val (x, y, z) = sides
        val isValid = x * y * z != 0.0 && x + y > z

        if (!isValid) throw IllegalArgumentException()
        return isValid
    }
}
